split csv lines as QStrings and return a list of all fields.

tsteven4 opened the pull request:
supply csv_linesplit to process lines as QStrings and return a QStringList of all the fields.  This eliminates some oscillation between QStrings and char strings and eliminates a common memory leak when users of csv_lineparse fail to finish parsing a line leaving csv_lineparse holding dynamic memory.

add a RFC4180 based quoting strategy for csv, and use it for unicsv.  I left the historic strategy in place for other users.

add a textstream class to simplify and centralize use of file based textstreams.

convert unicsv to textstreams and csv_linesplit.  add option for specifying a codec with unicsv input and output.  print available codecs when an unrecognized codec is requested.   add a testcase to demonstrate that we can enquote and dequote a unicsv field and get the original field back (which we didn't previously do).

convert ozi to use textstreams and csv_linesplit.  convert main to use csv_linesplit.

other formats continue to use the legacy character based csv_lineparse: compegps, g7towin, garmin_txt, gopal, jtr, maggeo, netstumbler, nmn4, pocketfms_wp, stmsdf, tmpro.  If usages justifies the investment we can convert these as well.

tsteven4 committed:
introdcue csv_linesplit

Which is like csv_lineparse, except it gives you a list
of all the values, and it processing the line as a QString.

tsteven4 committed:
Merge branch 'master' into stringsplit

tsteven4 committed:
introduce gpsbabel::TextStream, use it

in unicsv reader, ozi.

tsteven4 committed:
add missing new files.

tsteven4 committed:
convert unicsv writer to textstream.

tsteven4 committed:
if a codec is not found list available.

tsteven4 committed:
switch unicsv to CET_CHARSET_UTF8 to avoid

undesired fs conversions.  All conversions are
handled by the codec used by qtextstream.

tsteven4 committed:
output boms with non utf8 unicode codecs.

tsteven4 committed:
use rfc4180 dequote method with unicsv.

tsteven4 committed:
add test for csv quoting w RFC4180.

tsteven4 committed:
add missing reference files for csv test.

GPSBabelDeveloper commented on ozi.cc:
```diff
@@ -765,11 +743,7 @@ data_read()
-    if (buff.isNull()) {
-      break;
-    }
+  while (buff = stream->readLine(), !buff.isNull()) {
```
Isn't this just while (buff = stream->readLine()) { ?Shouldn't the next return be null anyway?

tsteven4 replied:
No.  buff is a QString, which cannot be automatically converted to a bool.

error: could not convert ‘buff.QString::operator=(QTextStream::readLine(qint64)(0))’ from ‘QString’ to ‘bool’

GPSBabelDeveloper commented on GPSBabel.pro:
```diff
@@ -152,6 +153,7 @@ HEADERS =  \
 	src/core/datetime.h \
 	src/core/file.h \
 	src/core/logging.h \
+  src/core/textstream.h \
```
Is there a tab here that's invisible to you?

tsteven4 replied:
expandtab got me again.  I fixed it and another line with the same issue.  We don't really need any tabs here at all in the pro file, but for consistency I put the tab on the textstream.h and xcsv.h lines.

GPSBabelDeveloper commented on ozi.cc:
```diff
@@ -267,7 +245,7 @@ ozi_openfile(const QString& fname)
 
   if (fname == "-") {
-    if (ozi_file.file == nullptr) {
+    if (stream == nullptr) {
```
I know this is mechanical, but I wonder if stream is ever !nullptr here.

tsteven4 replied:
yes, it can be depending on the number of waypoints, tracks and routes.

when fname == "-" then the first time here stream will be nullptr, but subsequent calls it will be != nullptr.

The first call possibility is from data_write() if we have one or more waypoints.
Subsequent call possibilities are from data_write() if we have tracks, one call per track header.
And a final possibility from data_write() if we have one or more routes.

GPSBabelDeveloper commented on unicsv.cc:
```diff
@@ -1228,15 +1204,15 @@ unicsv_parse_one_line(char* ibuf)
-  while ((buff = gbfgetstr(fin))) {
-    buff = lrtrim(buff);
-    if ((*buff == '\0') || (*buff == '#')) {
+  while ((buff = fin->readLine(), !buff.isNull())) {
```
As above.

tsteven4 replied:
same as above, no auto conversion from QString to bool.

GPSBabelDeveloper reviewed (approved):
Ambitious and a much needed change. Thank you. There are some cases where we should be using auto_ptr as our idiom, but when "just" doing mechanical transformations, it's hard to hold that as the bar.
Go for it.

tsteven4 committed:
use consistent indending in GPSBabel.pro

The HEADER assignment now consistenly uses tabs,
while everything else uses spaces!

tsteven4 committed:
appease clazy wrt container might detach.
